text_generation/causal_lm/cpp/continuous_batching/library/src/tokenizer.cpp

```diff
@@ -1,73 +0,0 @@
-
-// Copyright (C) 2023-2024 Intel Corporation
-// SPDX-License-Identifier: Apache-2.0
-
-#include <mutex>
-#include "openvino/runtime/core.hpp"
-
-#include "tokenizer.hpp"
-
-class Tokenizer::Impl {
-    const size_t TOKENIZER_BATCH_SIZE = 1;
-    ov::InferRequest m_tokenizer;
-    ov::InferRequest m_detokenizer;
-    std::size_t m_eos_token_id;
-    //Using multiple infer requests hangs. For now we synchronize entire execution on a single infer request.
-    std::mutex m_tokenizer_mutex;
-    std::mutex m_detokenizer_mutex;
-
-public:
-    explicit Impl(const std::string& models_path)
-    {
-        ov::Core core;
-        core.add_extension(OPENVINO_TOKENIZERS_PATH);  // OPENVINO_TOKENIZERS_PATH is defined in CMakeLists.txt
-
-        std::shared_ptr<ov::Model> tokenizer_model = core.read_model(models_path + "/openvino_tokenizer.xml");
-        const ov::AnyMap& rt_info = tokenizer_model->get_rt_info();
-        OPENVINO_ASSERT(rt_info.find("eos_token_id") != rt_info.end(), "Failed to detect \"eos_token_id\" in openvino_tokenizer.xml runtime information");
-        m_eos_token_id = rt_info.at("eos_token_id").as<int64_t>();
-
-        // tokenizer and detokenizer work on CPU only
-        m_tokenizer = core.compile_model(
-            tokenizer_model, "CPU").create_infer_request();
-        m_detokenizer = core.compile_model(
-            models_path + "/openvino_detokenizer.xml", "CPU").create_infer_request();
-    }
-
-    ov::Tensor encode(std::string prompt) {
-        std::unique_lock<std::mutex> lock(m_tokenizer_mutex);
-        m_tokenizer.set_input_tensor(ov::Tensor{ov::element::string, {TOKENIZER_BATCH_SIZE}, &prompt});
-        m_tokenizer.infer();
-        ov::Tensor tmp_tensor = m_tokenizer.get_tensor("input_ids");
-        ov::Tensor output_tensor(tmp_tensor.get_element_type(), tmp_tensor.get_shape());
-        tmp_tensor.copy_to(output_tensor);
-        return output_tensor;
-    }
-
-    std::string decode(std::vector<int64_t> tokens) {
-        std::unique_lock<std::mutex> lock(m_detokenizer_mutex);
```

Comment:
@Wovchena CB version of Tokenizer class had a mutex to allow usage from parallel threads
I suppose it's harmless to insert it in common version of Tokenizer? (client implementation)

CC @mzegla

Reply:
Yes, I took a look at tokenizer in in main genai sources. It looks very similar, but will not work if encode/decode is called in parallel from multiple threads which is the case in serving scenario (that's why we added synchronization here). 

I suppose the alternative could be to create new infer request for every inference or have some kind of infer requests pool. I didn't do that in the first place because it also had some issues and the whole pipeline hanged, but recently I didn't observe such behavior, so maybe that would be a direction worth considering.

Reply:
Done